### PR TITLE
feat: Allow hiding password auth, changing OpenID Connect text and OpenID Connect icon

### PR DESCRIPTION
**Pull request by normana10**
Just some ideas to help decrease friction in onboarding developers to my coder deployment

For context, my users will _never_ use password auth, only Oauth 

Examples:

1. Default coder login (nothing new here, just for reference)

2. With the following env vars set (also nothing new here, just for reference):
  - CODER_OIDC_ISSUER_URL
  - CODER_OIDC_EMAIL_DOMAIN
  - CODER_OIDC_CLIENT_ID
  - CODER_OIDC_CLIENT_SECRET

3. With the following env vars set:
  - CODER_OIDC_ISSUER_URL
  - CODER_OIDC_EMAIL_DOMAIN
  - CODER_OIDC_CLIENT_ID
  - CODER_OIDC_CLIENT_SECRET
  - CODER_PASSWORD_AUTH_HIDDEN=true
  - CODER_OIDC_SIGN_IN_TEXT="Sign in with Gitea"
  - CODER_OIDC_ICON_URL=https://gitea.io/images/gitea.png

I only reference Gitea because that's my OpenID Connect provider at home, feel free to replace that with anything you like 👍

I also acknowledge that I'm changing the shape of the return body of `api/v2/users/authmethods`. Not sure if that's a deal breaker?

Also let me know if Coder _intentionally_ didn't want to do any of these. Don't want my random PRs to steamroll any actual conversations that have taken place

### CODE REVIEW

**Review by kylecarbs — COMMENTED** [review]
Hey 👋 I appreciate the contribution!

I think we'll want to end up disabling password authentication instead of hiding it for security reasons. I'll review this later today!

**Comment by bpmct** [comment]
@kylecarbs and I just had a chat and decided we shouldn't completely disable the password login so the default `Owner` account can log in.

However, we can put it behind a backdoor (e.g. `?auth=password`) so that users don't see an option to log in with a user/pass. Also, OIDC users wouldn't be able to log in with a password anyways

**Comment by normana10** [comment]
Implemented

With the following env vars set:

- CODER_OIDC_ISSUER_URL
- CODER_OIDC_EMAIL_DOMAIN
- CODER_OIDC_CLIENT_ID
- CODER_OIDC_CLIENT_SECRET
- CODER_PASSWORD_AUTH_HIDDEN=true
- CODER_OIDC_SIGN_IN_TEXT="Sign in with Gitea"
- CODER_OIDC_ICON_URL=https://gitea.io/images/gitea.png

image
image

**Comment by mafredri** [comment]
> I think we'll want to end up disabling password authentication instead of hiding it for security reasons. I'll review this later today!

@kylecarbs how would an owner/admin sign in if we disable password auth? As @bpmct proposed, we could put it behind `?password=password`, or maybe better `/login/admin`. But ultimately that seems like an inconvenience. Maybe it's negligible though.

**Review by mafredri — COMMENTED** [review]
Backend changes look good to me with one minor request if we keep the flag.

**Comment by kylecarbs** [comment]
I think I'd prefer if we did a small "Login with Password" link at the bottom that would permit authenticating with a password. It's awkward to not allow it, since in emergency situations it would be required.

An alternative is to take the stance that owners/admins should be able to modify the deployment, so they could always disable this feature.

**Comment by bpmct** [comment]
> I think I'd prefer if we did a small "Login with Password" link at the bottom that would permit authenticating with a password. It's awkward to not allow it, since in emergency situations it would be required.
> 
> 
> 
> An alternative is to take the stance that owners/admins should be able to modify the deployment, so they could always disable this feature.

I can get behind a small link as long as it's clear that SSO is preferred. Many login forms tend to have both even if 99% of an organization's users use SSO

File: cli/deployment/config.go [hunk 1]
```diff
@@ -158,6 +158,11 @@ func newConfig() *codersdk.DeploymentConfig {
 			Flag:   "postgres-url",
 			Secret: true,
 		},
+		PasswordAuthHidden: &codersdk.DeploymentConfigField[bool]{
+			Name:  "Flag to hide password auth",
+			Usage: "When this flag is set to true, the user/password form in the UI will be hidden",
+			Flag:  "password-auth-hidden",
```
[the remark this answers is not in the document]
Comment (normana10) [reply cli/deployment/config.go:164]: Makes sense to me (will be included in next push)

**Comment by normana10** [comment]
@kylecarbs I agree with it being awkward and can add a little link to present the username/password field

As @mafredri mentions, would you like this to _just happen_ if OIDC is configured? Then I can remove the "hide-password-auth" flag I added in here

Either way, let me know and I'll go implement

Sorry for the delay, busy week and #Holidays

**Comment by kylecarbs** [comment]
I think this can _just happen_ if OIDC is configured, yes!

Sorry for the delays on my end too - holidays!

**Comment by normana10** [comment]
Ok this now looks like (default password-only):
PasswordOnly

With the following env vars set:

CODER_OIDC_ISSUER_URL
CODER_OIDC_EMAIL_DOMAIN
CODER_OIDC_CLIENT_ID
CODER_OIDC_CLIENT_SECRET
CODER_PASSWORD_AUTH_HIDDEN=true
CODER_OIDC_SIGN_IN_TEXT="Sign in with Gitea"
CODER_OIDC_ICON_URL=https://gitea.io/images/gitea.png

OIDC_Enabled

On clicking "Show password login" (Looks the way it does today):

OIDC_Enabled_Show_Password

File: codersdk/users.go [hunk 2]
```diff
@@ -105,11 +105,21 @@ type CreateOrganizationRequest struct {
...
+	Password AuthMethod `json:"password"`
+	Github   AuthMethod `json:"github"`
+	OIDC     OIDCMethod `json:"oidc"`
+}
+
+type AuthMethod struct {
+	Enabled bool `json:"enabled"`
+}
+
+type OIDCMethod struct {
```
Comment (kylecarbs) [inline codersdk/users.go:119]: Let's call this `OIDCAuthMethod` instead!

File: cli/deployment/config.go [hunk 3]
```diff
@@ -254,6 +254,16 @@ func newConfig() *codersdk.DeploymentConfig {
 				Flag:    "oidc-username-field",
 				Default: "preferred_username",
 			},
+			SignInText: &codersdk.DeploymentConfigField[string]{
+				Name:  "OpenID Connect sign in text",
+				Usage: "The text to show on the OpenID Connect sign in button",
+				Flag:  "oidc-sign-in-text",
```
Comment (kylecarbs) [inline cli/deployment/config.go:260]: I think defaulting to the OIDC sign-in text would be good here!

**Review by kylecarbs — APPROVED** [review]
Looks good, and thank you for the contribution 🥳

File: site/src/components/SignInForm/SignInForm.tsx [hunk 4]
```diff
@@ -143,98 +156,118 @@ export const SignInForm: FC<React.PropsWithChildren<SignInFormProps>> = ({
...
+            </Link>
+          )}
+        </Box>
+      )}
+      {!showPasswordAuth && (
+        <Typography
+          className={styles.showPasswordLink}
+          onClick={() => setShowPasswordAuth(true)}
+        >
+          Show password login
```
Comment (Kira-Pilot) [inline site/src/components/SignInForm/SignInForm.tsx:269]: Do you mind adding a translation here?

File: site/src/pages/LoginPage/LoginPage.test.tsx [hunk 5]
```diff
@@ -10,6 +10,7 @@ import {
 } from "../../testHelpers/renderHelpers"
 import { server } from "../../testHelpers/server"
 import { LoginPage } from "./LoginPage"
+import * as TypesGen from "../../api/typesGenerated"
```
Comment (Kira-Pilot) [inline site/src/pages/LoginPage/LoginPage.test.tsx:13]:
```suggestion
import * as TypesGen from "api/typesGenerated"
```

**Review by bpmct — APPROVED** [review]
This UX LGTM! Thanks for the contribution, let's ship it 🚢

File: site/src/pages/LoginPage/LoginPage.test.tsx [hunk 6]
```diff
@@ -120,4 +123,34 @@ describe("LoginPage", () => {
...
+      await screen.queryByText(Language.passwordSignIn),
+    ).not.toBeInTheDocument()
+    await screen.findByText(Language.githubSignIn)
+
+    const showPasswordAuthLink = screen.getByText("Show password login")
+    await userEvent.click(showPasswordAuthLink)
+
+    await screen.findByText(Language.passwordSignIn)
+    await screen.findByText(Language.githubSignIn)
+  })
```
Comment (Kira-Pilot) [inline site/src/pages/LoginPage/LoginPage.test.tsx:155]: Thanks for these tests!

**Review by Kira-Pilot — APPROVED** [review]

File: site/src/components/SignInForm/SignInForm.tsx [hunk 7]
```diff
@@ -143,98 +156,118 @@ export const SignInForm: FC<React.PropsWithChildren<SignInFormProps>> = ({
...
-            fullWidth
-            id="password"
-            label={Language.passwordLabel}
-            type="password"
-            variant="outlined"
-          />
-          <div>
-            <LoadingButton
-              loading={isLoading}
+      {showPasswordAuth && (
```
[the remark this answers is not in the document]
Comment (normana10) [reply site/src/components/SignInForm/SignInForm.tsx:159]: I agree

I took a hack at it (might have made a few stylistic choices, open to feedback)

Also turned these `var &&` into `<Maybe`s as I saw that used all over the codebase (and might have to steal that for myself)

**Comment by normana10** [comment]
Oops sorry pushed the golden file update from a different computer

File: site/src/components/SignInForm/SignInForm.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kira-Pilot) [reply site/src/components/SignInForm/SignInForm.tsx]: Looks beautiful! Thanks for leaving things better than you found them :)

**Comment by kylecarbs** [comment]
@normana10 seems like `make gen` isn't quite producing the same for you... want me to post a patch you can apply?

**Comment by normana10** [comment]
@kylecarbs 

Feel free

I keep trying to run `make gen` and it does nothing

Then I do a `make gen -B`

And I get a:
```
$ make gen -B
panic: could not start resource: : dial unix /var/run/docker.sock: connect: connection refused

goroutine 1 [running]:
main.main()
	/Users/normana10/git/coder/coderd/database/gen/dump/main.go:23 +0xb25
exit status 2
make: *** [Makefile:461: coderd/database/dump.sql] Error 1
```

(On my Mac)

Which.... totally makes sense, I don't have Docker installed on my Mac

Buuuut, when I try the same on my Fedora computer it ends up erroring out with a:
```
...
../docs/api/workspaces.md 99ms
../docs/manifest.json 16ms
../coderd/apidoc/swagger.json 351ms
Done in 4.37s.
ERROR: The 'yq' dependency is required, but is not available.

ERROR: One or more dependencies are not available, check above log output for more details.
make: *** [Makefile:509: site/.prettierrc.yaml] Error 1
```

But... `yq` is definitely installed

**Comment by mafredri** [comment]
@normana10 sorry the error is so unhelpful, you'll need to have `yq` v4 installed, either as `yq4` or `yq --version >= v4`.

**Comment by normana10** [comment]
```
$ yq --version
yq version 4.2.0
```

🤷

**Comment by normana10** [comment]
Wait I just noticed that `yq` _outside_ of nix-shell is v4 but _inside_ it's v3 o.O

Just symlinked my install of `yq` to `yq4` and `make gen -B` ran just fine

Pushing up now

**Comment by normana10** [comment]
That was a roller coaster but I think it's all good now? 👍

Thanks!

**Comment by kylecarbs** [comment]
Woot woot. Thanks for pushing through @normana10 :partying_face:

**Comment by kylecarbs** [comment]
@normana10 can you merge in `main` then I'll merge this in ASAP! Thank you again for being patient, this was a controversial one for us.

**Comment by normana10** [comment]
@kylecarbs 

Done

Sorry if I caused any issues 😬

**Comment by kylecarbs** [comment]
@normana10 we're now using this on https://dev.coder.com :sunglasses: 

image

**Pull request merged**